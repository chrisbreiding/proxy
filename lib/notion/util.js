const rp = require('request-promise')

const makeRequest = ({ notionToken, body, method = 'get', path }) => {
  return rp({
    method,
    body,
    uri: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

const getBlockChildren = ({ notionToken, pageId }) => {
  return makeRequest({ notionToken, path: `blocks/${pageId}/children` })
}

const appendBlockChildren = ({ notionToken, pageId, blocks }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
}

const updateBlock = ({ notionToken, blockId, block }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: block,
  })
}

const getPlainText = (withText) => {
  if (!withText || !withText.text) return

  return withText.text
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

const weatherIconMap = {
  'clear-day': '☀️',
  'clear-night': '☀️',
  'rain': '☔️',
  'snow': '❄️',
  'sleet': '🌨',
  'wind': '💨',
  'fog': '🌫',
  'cloudy': '☁️',
  'partly-cloudy-day': '⛅️',
  'partly-cloudy-night': '⛅️',
  'default': '🌑',
}

const getWeatherIcon = (iconName) => {
  return weatherIconMap[iconName] || weatherIconMap.default
}

const makeTextPart = (content, color) => {
  const textPart = { text: { content } }

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

module.exports = {
  getBlockChildren,
  appendBlockChildren,
  updateBlock,
  getPlainText,
  getWeatherIcon,
  makeTextPart,
}
