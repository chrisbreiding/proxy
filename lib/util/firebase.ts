import admin, { firestore } from 'firebase-admin'
import { readJsonSync } from 'fs-extra'
import path from 'path'

import { getEnv } from './env'
import { basePath } from './persistent-data'

const apps = {
  tv: {
    credentialsPath: 'firebase-tv-credentials.json',
    databaseURL: getEnv('FIREBASE_TV_DATABASE_URL'),
  },
  sync: {
    credentialsPath: 'firebase-sync-credentials.json',
    databaseURL: getEnv('FIREBASE_SYNC_DATABASE_URL'),
  },
}

export function initializeApp (appName: keyof typeof apps) {
  const app = apps[appName]

  if (!app) {
    throw new Error(`Unknown app: ${appName}`)
  }

  const serviceAccount = readJsonSync(path.join(basePath, app.credentialsPath), { throws: false })

  // this should only happen when testing
  if (!serviceAccount) {
    throw new Error(`No service account found for ${appName}`)
  }

  admin.initializeApp({
    credential: admin.credential.cert(serviceAccount),
    databaseURL: app.databaseURL,
  })

  const db = admin.firestore()

  db.settings({ ignoreUndefinedProperties: true })

  return db
}

type Snapshot = admin.firestore.QuerySnapshot<admin.firestore.DocumentData>

function getDataFromSnapshot<T> (snapshot: Snapshot): Promise<T[]> {
  return Promise.all(snapshot.docs.map(async (docSnapshot) => {
    const doc = await docSnapshot.ref.get()

    return Object.assign(doc.data()!, {
      id: doc.id,
    }) as T
  }))
}

export async function getCollection<T> (db: firestore.Firestore, collectionName: string): Promise<T[]> {
  const snapshot = await db.collection(collectionName).get()

  return getDataFromSnapshot<T>(snapshot)
}

interface Identifiable {
  id: string
}

export async function getSubCollections<T extends Identifiable, U> (db: firestore.Firestore, data: T[], collection1Name: string, collection2Name: string): Promise<U[]> {
  return Promise.all(data.map(async (datum: T) => {
    const snapshot = await db.collection(`${collection1Name}/${datum.id}/${collection2Name}`).get()

    const collections = await Promise.all(snapshot.docs.map(async (docSnapshot) => {
      return (await docSnapshot.ref.get()).data()!
    }))

    return Object.assign(datum, {
      [collection2Name]: collections,
    }) as U
  }))
}

export async function getDoc<T> (db: firestore.Firestore, docPath: string): Promise<T | undefined> {
  const doc = await db.doc(docPath).get()

  return doc.data() as T
}

type Condition = [string, admin.firestore.WhereFilterOp, any]

export async function getDocWhere<T> (db: firestore.Firestore, collectionName: string, condition: Condition): Promise<T | undefined> {
  const [fieldPath, opStr, value] = condition
  const snapshot = await db.collection(collectionName).where(fieldPath, opStr, value).get()

  if (snapshot.empty) {
    return
  }

  return (await getDataFromSnapshot<T>(snapshot))[0]
}

export async function addDoc (db: firestore.Firestore, docPath: string, value: any) {
  await db.doc(docPath).set(value)
}

export async function setDoc (db: firestore.Firestore, docPath: string, value: any) {
  await db.doc(docPath).set(value)
}

export async function updateDoc (db: firestore.Firestore, docPath: string, value: any) {
  await db.doc(docPath).update(value)
}

export async function deleteDoc (db: firestore.Firestore, docPath: string) {
  await db.doc(docPath).delete()
}
