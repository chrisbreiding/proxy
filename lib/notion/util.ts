import type { Method } from 'axios'
import dayjs from 'dayjs'
import type {
  ApiColor,
  BlockObjectResponse,
  BookmarkBlockObjectResponse,
  BulletedListItemBlockObjectResponse,
  CalloutBlockObjectResponse,
  ChildDatabaseBlockObjectResponse,
  ChildPageBlockObjectResponse,
  ColumnBlockObjectResponse,
  ColumnListBlockObjectResponse,
  DividerBlockObjectResponse,
  EmbedBlockObjectResponse,
  EquationBlockObjectResponse,
  FileBlockObjectResponse,
  Heading1BlockObjectResponse,
  Heading2BlockObjectResponse,
  Heading3BlockObjectResponse,
  ImageBlockObjectResponse,
  LinkPreviewBlockObjectResponse,
  LinkToPageBlockObjectResponse,
  ListBlockChildrenResponse,
  ListDatabasesResponse,
  NumberedListItemBlockObjectResponse,
  ParagraphBlockObjectResponse,
  PdfBlockObjectResponse,
  QuoteBlockObjectResponse,
  RichTextItemResponse,
  SyncedBlockBlockObjectResponse,
  TableBlockObjectResponse,
  TableOfContentsBlockObjectResponse,
  TableRowBlockObjectResponse,
  TemplateBlockObjectResponse,
  ToDoBlockObjectResponse,
  ToggleBlockObjectResponse,
  UnsupportedBlockObjectResponse,
  UpdateBlockBodyParameters,
  VideoBlockObjectResponse,
} from '@notionhq/client/build/src/api-endpoints'

import { clone, compact } from '../util/collections'
import { debug } from '../util/debug'
import { request } from '../util/network'

export const dateRegex = /(?:Sun|Mon|Tue|Wed|Thu|Fri|Sat), (\d{1,2}\/\d{1,2})/

export function getDateFromText (dateText: string) {
  const currentDate = dayjs()
  // originally assume the date's year matches the current year
  const date = dayjs(`${dateText}/${currentDate.year()}`, 'M/D/YYYY')
  // if the current month is after the date's month, we've crossed over years
  // and the date's year should be the next year. usually, it's because it's
  // currently December, but the date's month is January. this can be
  // generically applied, so it will work if it's currently June, but we're
  // trying to get the weather for April. we don't want past weather, only
  // future weather, so assume it's for the following year.
  if (currentDate.month() > date.month()) {
    return date.add(1, 'year')
  }

  return date
}

interface MakeRequestOptions {
  body?: object
  method?: Method
  notionToken: string
  path: string
}

function makeRequest<T> (options: MakeRequestOptions): Promise<T> {
  const { notionToken, body, path, method = 'get' } = options

  return request({
    method,
    body,
    url: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2022-06-28',
    },
  })
}

interface GetBlockChildrenOptions {
  notionToken: string
  pageId: string
}

export async function getBlockChildren ({ notionToken, pageId }: GetBlockChildrenOptions) {
  const { results } = await makeRequest<ListBlockChildrenResponse>({ notionToken, path: `blocks/${pageId}/children` }) as { results: BlockObjectResponse[] }

  return results.map(convertNotionBlockToOwnBlock)
}

interface GetBlockChildrenDeepOptions {
  filter?: (block: NotionBlock) => boolean
  includeId?: boolean
  notionToken: string
  pageId: string
}

export async function getBlockChildrenDeep ({ notionToken, pageId, filter, includeId }: GetBlockChildrenDeepOptions): Promise<NotionBlock[]> {
  const blocks = await getBlockChildren({ notionToken, pageId })
  const filteredBlocks = filter ? blocks.filter(filter) : blocks
  const blocksWithChildren = await Promise.all(filteredBlocks.map(async (block) => {
    if (block.has_children && block.type !== 'child_page') {
      block.children = await getBlockChildrenDeep({ notionToken, pageId: block.id, includeId })
    }

    return block
  }))

  return compact(blocksWithChildren)
}

type Content = BookmarkBlockObjectResponse['bookmark']
| BulletedListItemBlockObjectResponse['bulleted_list_item']
| CalloutBlockObjectResponse['callout']
| ChildDatabaseBlockObjectResponse['child_database']
| ChildPageBlockObjectResponse['child_page']
| ColumnBlockObjectResponse['column']
| ColumnListBlockObjectResponse['column_list']
| DividerBlockObjectResponse['divider']
| EmbedBlockObjectResponse['embed']
| EquationBlockObjectResponse['equation']
| FileBlockObjectResponse['file']
| Heading1BlockObjectResponse['heading_1']
| Heading2BlockObjectResponse['heading_2']
| Heading3BlockObjectResponse['heading_3']
| ImageBlockObjectResponse['image']
| LinkPreviewBlockObjectResponse['link_preview']
| LinkToPageBlockObjectResponse['link_to_page']
| NumberedListItemBlockObjectResponse['numbered_list_item']
| ParagraphBlockObjectResponse['paragraph']
| PdfBlockObjectResponse['pdf']
| QuoteBlockObjectResponse['quote']
| SyncedBlockBlockObjectResponse['synced_block']
| TableBlockObjectResponse['table']
| TableOfContentsBlockObjectResponse['table_of_contents']
| TableRowBlockObjectResponse['table_row']
| TemplateBlockObjectResponse['template']
| ToDoBlockObjectResponse['to_do']
| ToggleBlockObjectResponse['toggle']
| UnsupportedBlockObjectResponse['unsupported']
| VideoBlockObjectResponse['video']

export interface OwnBlock {
  children?: OwnBlock[] | NotionBlock[]
  type: BlockObjectResponse['type']
  content: Content
}

export interface NotionBlock extends OwnBlock {
  children?: NotionBlock[]
  has_children: boolean
  id: string
}

export type Block = OwnBlock | NotionBlock

function getRichText (block: OwnBlock): RichTextItemResponse[] | undefined {
  return 'rich_text' in block.content ? block.content.rich_text : undefined
}

interface MakeAppendRequestOptions {
  blocks: OutgoingBlock[]
  notionToken: string
  pageId: string
}

export function makeAppendRequest ({ notionToken, pageId, blocks }: MakeAppendRequestOptions) {
  return makeRequest<ListBlockChildrenResponse>({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface AppendBlockChildrenOptions {
  blocks: OwnBlock[]
  notionToken: string
  pageId: string
}

export async function appendBlockChildrenDeep ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  function moveChildren (blocks: NotionBlock[] | OwnBlock[]) {
    return blocks.map((block) => {
      const convertedBlock = convertBlockToOutgoingBlock(block)

      if (!block.children) return convertedBlock

      // @ts-ignore
      convertedBlock[block.type].children = moveChildren(block.children)

      return convertedBlock
    })
  }

  return makeAppendRequest({ notionToken, pageId, blocks: moveChildren(blocks) })
}

export async function appendBlockChildren ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  let toAppend: OutgoingBlock[] = []

  for (const block of blocks) {
    const convertedBlock = convertBlockToOutgoingBlock(block)

    if (block.children) {
      const children = block.children

      toAppend.push(convertedBlock)

      const { results } = await makeAppendRequest({ notionToken, pageId, blocks: toAppend })
      toAppend = []

      await appendBlockChildren({
        notionToken,
        pageId: results[results.length - 1].id,
        blocks: children,
      })
    } else {
      toAppend.push(convertedBlock)
    }
  }

  if (toAppend.length) {
    return makeAppendRequest({
      notionToken,
      pageId,
      blocks: toAppend,
    })
  }
}

interface UpdateBlockOptions {
  notionToken: string
  blockId: string
  block: Block
}

export function updateBlock ({ notionToken, blockId, block }: UpdateBlockOptions) {
  const convertedBlock = convertBlockToOutgoingBlock(block)

  return makeRequest<void>({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: convertedBlock,
  })
}

interface UpdatePageOptions {
  notionToken: string
  pageId: string
  properties: {
    [key: string]: {
      date: {
        start: string
      }
    }
  }
}

export function updatePage ({ notionToken, pageId, properties }: UpdatePageOptions) {
  return makeRequest<void>({
    notionToken,
    body: { properties },
    method: 'patch',
    path: `pages/${pageId}`,
  })
}

interface QueryDatabasesOptions {
  notionToken: string
  databaseId: string
}

export function queryDatabases ({ notionToken, databaseId }: QueryDatabasesOptions) {
  return makeRequest<ListDatabasesResponse>({
    notionToken,
    method: 'post',
    path: `databases/${databaseId}/query`,
  })
}

export function convertNotionBlockToOwnBlock (block: BlockObjectResponse): NotionBlock {
  const type = block.type

  return {
    // @ts-ignore
    content: clone(block[type] as Content),
    has_children: block.has_children,
    id: block.id,
    type,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

type OutgoingBlock = UpdateBlockBodyParameters & {
  object: string
  type: BlockObjectResponse['type']
}

export function convertBlockToOutgoingBlock (block: Block): OutgoingBlock {
  return {
    object: 'block',
    type: block.type,
    [block.type]: block.content,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface MakeBlockOptions {
  annotations?: Partial<RichTextItemResponse['annotations']>
  children?: OwnBlock[] | NotionBlock[]
  text: string
  type: BlockObjectResponse['type']
}

export function makeBlock ({ text, type = 'paragraph', annotations, children }: MakeBlockOptions) {
  return {
    type,
    children,
    content: {
      rich_text: [
        {
          type: 'text',
          text: {
            content: text,
          },
          annotations,
          plain_text: text,
        },
      ],
    },
  } as OwnBlock
}

export function richTextToPlainText (richText: RichTextItemResponse[]) {
  return richText
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

export function getBlockPlainText (block: Block) {
  const richText = getRichText(block)

  if (!richText) return

  return richTextToPlainText(richText)
}

interface RichTextItem {
  text: {
    content: string
    link?: { url: string } | null
  }
  type?: 'text'
  annotations?: {
    bold?: boolean
    italic?: boolean
    strikethrough?: boolean
    underline?: boolean
    code?: boolean
    color?: ApiColor
  }
}

export function makeTextPart (content: string, color?: ApiColor) {
  const textPart = { text: { content } } as RichTextItem

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

export function getMonthNameFromIndex (monthIndex: number, short = false) {
  return dayjs().month(monthIndex).format(short ? 'MMM' : 'MMMM')
}

export function getMonths ({ short }: { short?: boolean } = {}) {
  return Array.from(new Array(12)).map((_, i) => {
    return getMonthNameFromIndex(i, short)
  })
}

/* c8 ignore start */
const getDisplayPrefix = (block: NotionBlock) => {
  switch (block.type) {
    case 'paragraph': return ''
    case 'heading_1': return '# '
    case 'heading_2': return '## '
    case 'heading_3': return '### '
    case 'to_do': return `[${'checked' in block.content && block.content.checked ? '✓' : ' '}] `
    case 'bulleted_list_item': return '• '
    case 'numbered_list_item': return '1. '
    case 'toggle': return '> '
    case 'code': return '<> '
    case 'quote': return '| '
    case 'callout': return '" '
    default: return ''
  }
}

function getDisplayText (block: NotionBlock) {
  if (block.type === 'child_page') {
    return `[${'title' in block.content ? block.content.title : '<untitled>'}]`
  }

  if (block.type === 'divider') {
    return '-----'
  }

  if (getRichText(block)) {
    return `${getDisplayPrefix(block)}${getBlockPlainText(block)}`
  }

  return `(${block.type})`
}

// displays block for logging/debugging purposes
export function displayBlocks (blocks: NotionBlock[]) {
  blocks.forEach((block) => {
    debug(`[${block.id}] ${getDisplayText(block)}${block.has_children ? ' (parent)' : ''}`)
  })
}
/* c8 ignore stop */
