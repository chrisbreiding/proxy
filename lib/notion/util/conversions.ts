import type { BlockObjectResponse, RichTextItemResponse } from '@notionhq/client/build/src/api-endpoints'
import type { Block, NotionBlock, OutgoingBlock } from '../types'
import { clone } from '../../util/collections'

export function convertNotionBlockToOwnBlock (block: BlockObjectResponse): NotionBlock {
  const type = block.type
  // @ts-ignore
  const content = block[type] as Content

  return {
    content: clone(content),
    has_children: block.has_children,
    id: block.id,
    type,
  }
}

export function convertBlockToOutgoingBlock (block: Block): OutgoingBlock {
  return {
    object: 'block',
    type: block.type,
    [block.type]: block.content,
  }
}

export function richTextToPlainText (richText: RichTextItemResponse[]) {
  return richText
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}
