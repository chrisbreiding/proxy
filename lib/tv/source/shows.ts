import dayjs from 'dayjs'

import { makeRequest } from './util'
import { debug, debugVerbose } from '../../util/debug'

interface TheTvDbShow {
  objectID: string
  aliases: string[]
  country: string
  id: string // "series-81189"
  image_url: string // "https://artworks.thetvdb.com/banners/posters/81189-10.jpg"
  name: string // "Breaking Bad"
  first_air_time: string // "2008-01-20",
  overview: string // "When Walter White, a chemistry teacher, is...",
  primary_language: string
  primary_type: string
  status: 'Continuing' | 'Ended'
  type: string
  tvdb_id: string // "81189"
  year: string
  slug: string
  overviews: {
    [key: string]: string
  },
  translations: {
    [key: string]: string
  },
  network: string // "AMC"
  remote_ids: {
    id: string
    type: number
    sourceName: string
  }[]
  thumbnail: string // "https://artworks.thetvdb.com/banners/posters/81189-10_t.jpg"
}

export interface SearchResultShow {
  description: string
  firstAired?: string
  id: string
  name: string
  network: string
  poster?: string
  status: TheTvDbShow['status']
}

function convert (show: TheTvDbShow): SearchResultShow {
  return {
    description: show.overview,
    firstAired: show.first_air_time ? dayjs(show.first_air_time).toISOString() : undefined,
    id: `${show.tvdb_id}`,
    name: show.name,
    network: show.network,
    poster: show.image_url,
    status: show.status,
  }
}

export async function searchShows (showName: string): Promise<SearchResultShow[]> {
  debugVerbose('search for', showName)

  try {
    const { data: shows } = await makeRequest({
      path: 'search',
      params: {
        query: showName,
        type: 'series',
      },
    })

    return shows.map(convert)
  } catch (error: any) {
    debug(`Searching ${showName} failed:`, error?.stack || error)

    throw error
  }
}

interface TheTvDbShowUpdate {
  recordType: string
  recordId: number // 83498
  methodInt: number
  method: string
  extraInfo: string
  userId: number
  timeStamp: number // 1667707258
  entityType: string // "series"
}

interface ShowUpdate {
  id: string
  dateTime: string
}

function convertShowUpdates (showUpdate: TheTvDbShowUpdate): ShowUpdate {
  return {
    id: `${showUpdate.recordId}`,
    dateTime: dayjs.unix(showUpdate.timeStamp).toISOString(),
  }
}

export async function getShowsUpdatedSince (date: string): Promise<ShowUpdate[]> {
  debugVerbose('find shows updated since', date)

  try {
    const { data: showUpdates } = await makeRequest({
      path: 'updates',
      params: {
        action: 'update',
        type: 'series',
        since: `${dayjs(date).unix()}`,
      },
    })

    return showUpdates.map(convertShowUpdates)
  } catch (error: any) {
    debug(`Getting shows updated since ${date} failed:`, error?.stack || error)

    throw error
  }
}
