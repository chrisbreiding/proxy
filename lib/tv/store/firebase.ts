import admin from 'firebase-admin'
import { readJsonSync } from 'fs-extra'
import path from 'path'

import { getEnv } from '../../util/env'
import { basePath } from '../../util/persistent-data'

function initializeApp () {
  const serviceAccount = readJsonSync(path.join(basePath, 'firebase-tv-credentials.json'), { throws: false })

  // this should only happen when testing
  if (!serviceAccount) return

  admin.initializeApp({
    credential: admin.credential.cert(serviceAccount),
    databaseURL: getEnv('FIREBASE_TV_DATABASE_URL'),
  })

  const db = admin.firestore()

  db.settings({ ignoreUndefinedProperties: true })

  return db
}

const db = initializeApp()

type Snapshot = admin.firestore.QuerySnapshot<admin.firestore.DocumentData>

function getDataFromSnapshot<T> (snapshot: Snapshot): Promise<T[]> {
  return Promise.all(snapshot.docs.map(async (docSnapshot) => {
    const doc = await docSnapshot.ref.get()

    return Object.assign(doc.data()!, {
      id: doc.id,
    }) as T
  }))
}

export async function getCollection<T> (collectionName: string): Promise<T[]> {
  if (!db) return []

  const snapshot = await db.collection(collectionName).get()

  return getDataFromSnapshot<T>(snapshot)
}

interface Identifiable {
  id: string
}

export async function getSubCollections<T extends Identifiable, U> (data: T[], collection1Name: string, collection2Name: string): Promise<U[]> {
  if (!db) return []

  return Promise.all(data.map(async (datum: T) => {
    const snapshot = await db.collection(`${collection1Name}/${datum.id}/${collection2Name}`).get()

    const collections = await Promise.all(snapshot.docs.map(async (docSnapshot) => {
      return (await docSnapshot.ref.get()).data()!
    }))

    return Object.assign(datum, {
      [collection2Name]: collections,
    }) as U
  }))
}

export async function getDoc<T> (docPath: string): Promise<T | undefined> {
  if (!db) return

  const doc = await db.doc(docPath).get()

  return doc.data() as T
}

type Condition = [string, admin.firestore.WhereFilterOp, any]

export async function getDocWhere<T> (collectionName: string, condition: Condition): Promise<T | undefined> {
  if (!db) return

  const [fieldPath, opStr, value] = condition
  const snapshot = await db.collection(collectionName).where(fieldPath, opStr, value).get()

  if (snapshot.empty) {
    return
  }

  return (await getDataFromSnapshot<T>(snapshot))[0]
}

export async function addDoc (docPath: string, value: any) {
  if (!db) return

  await db.doc(docPath).set(value)
}

export async function addCollectionToDoc (collectionPath: string, values: any[]) {
  if (!db) return

  const batch = db.batch()

  values.forEach((value) => {
    const docRef = db.collection(collectionPath).doc(value.id)
    batch.set(docRef, value)
  })

  await batch.commit()
}

export async function updateDoc (docPath: string, value: any) {
  if (!db) return

  await db.doc(docPath).update(value)
}

export async function deleteDoc (docPath: string) {
  if (!db) return

  await db.doc(docPath).delete()
}
