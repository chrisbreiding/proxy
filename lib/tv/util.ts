import type { Method } from 'axios'
import { debug, debugVerbose } from '../util/debug'
import { getEnv } from '../util/env'
import { request } from '../util/network'

const apikey = getEnv('THETVDB_API_KEY')
const pin = getEnv('THETVDB_PIN')
export const baseUrl = 'https://api4.thetvdb.com'

interface LoginResult {
  status: string
  data: {
    token: string
  }
}

async function authenticate () {
  debugVerbose('authenticate')

  try {
    const { data, status } = await request({
      method: 'post',
      url: `${baseUrl}/v4/login`,
      body: { apikey, pin },
    }) as LoginResult

    if (!data?.token) {
      throw new Error(`Could not authenticate with TheTVDB, status: ${status}`)
    }

    return data.token as string
  } catch (error: any) {
    debug('Authentication failed:', error?.stack || error)

    throw error
  }
}

interface MakeRequestOptions {
  method?: Method
  params?: { [key: string]: string }
  path: string
  token?: string
}

export async function makeRequest (options: MakeRequestOptions) {
  const { method, params, path } = options
  const token = options.token || await authenticate()
  const url = `${baseUrl}/v4/${path}`

  debugVerbose('request: %o', { method, url, params, token })

  // TODO: handle error
  const result = await request({
    headers: {
      'Authorization': `Bearer ${token}`,
    },
    method,
    params,
    url,
  })

  result.token = token

  return result
}
