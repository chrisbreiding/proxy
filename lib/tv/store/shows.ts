import { Episode, getEpisodesForShow } from '../source/episodes'
import type { SearchResultShow } from '../source/shows'
import {
  addCollectionToDoc,
  addDoc,
  deleteDoc,
  getCollection,
  getDoc,
  updateDoc,
} from './firebase'
import type { User } from './users'

interface UserShow {
  displayName: string
  episodes: Episode[]
  fileName: string
  id: string
  poster?: string
  searchName: string
  status: SearchResultShow['status']
}

interface EditableShowProps {
  displayName: string
  fileName: string
  searchName: string
}

interface ShowProps {
  id: string
  name: string
  poster?: string
  status: SearchResultShow['status']
  users: {
    [key: string]: EditableShowProps
  }
}

interface FullShowProps extends ShowProps {
  episodes: Episode[]
}

export class Show implements ShowProps {
  id: string
  name: string
  poster?: string
  status: SearchResultShow['status']
  users: {
    [key: string]: EditableShowProps
  }

  static fromSourceShow (sourceShow: SearchResultShow, user: User) {
    return new Show({
      id: sourceShow.id,
      name: sourceShow.name,
      poster: sourceShow.poster,
      status: sourceShow.status,
      users: {
        [user.id]: {
          displayName: sourceShow.name,
          fileName: sourceShow.name,
          searchName: sourceShow.name,
        },
      },
    })
  }

  static forUser (showData: FullShowProps, user: User): UserShow {
    const userData = showData.users[user.id] || {}

    return {
      episodes: showData.episodes,
      id: showData.id,
      poster: showData.poster,
      status: showData.status,
      displayName: userData.displayName,
      fileName: userData.fileName,
      searchName: userData.searchName,
    }
  }

  constructor (props: ShowProps) {
    this.id = props.id
    this.name = props.name
    this.poster = props.poster
    this.status = props.status
    this.users = props.users
  }

  serialize () {
    return {
      id: this.id,
      name: this.name,
      poster: this.poster,
      status: this.status,
      users: this.users,
    }
  }

  addUser (user: User) {
    this.users[user.id] = {
      displayName: this.name,
      fileName: this.name,
      searchName: this.name,
    }
  }

  updateUser (user: User, props: EditableShowProps) {
    this.users[user.id] = props
  }
}

export async function getShows (user: User): Promise<UserShow[]> {
  const showData = await getCollection<FullShowProps>('shows')

  // TODO: optimize so that all episodes aren't
  // retrieved before getting episodes

  return showData
  .filter((showDatum) => {
    return !!showDatum.users[user.id]
  })
  .map((showDatum) => {
    return Show.forUser(showDatum, user)
  })
}

export async function addShow (sourceShow: SearchResultShow, user: User): Promise<Show | undefined> {
  const showDatum = await getDoc<ShowProps>(`shows/${sourceShow.id}`)

  // show does not exist in firebase
  // get it from TheTVDB, save it, and add to user
  if (!showDatum) {
    const episodes = await getEpisodesForShow(sourceShow.id)
    const show = Show.fromSourceShow(sourceShow, user)

    await addDoc(`shows/${show.id}`, show.serialize())
    await addCollectionToDoc(`shows/${show.id}/episodes`, episodes)

    return show
  }

  const show = new Show(showDatum)

  // show already has user
  if (show.users[user.id]) {
    return
  }

  // show exists in firebase, add user to it
  show.addUser(user)
  await updateDoc(`shows/${show.id}`, { users: show.users })

  return show
}

export async function updateShow (id: string, showUpdate: EditableShowProps, user: User) {
  const showDatum = await getDoc<ShowProps>(`shows/${id}`)

  if (!showDatum) return

  const show = new Show(showDatum)

  show.updateUser(user, showUpdate)
  await updateDoc(`shows/${id}`, show)

  return show
}

export async function deleteShow (id: string, user: User) {
  const showDatum = await getDoc<ShowProps>(`shows/${id}`)

  if (!showDatum?.users[user.id]) {
    return
  }

  delete showDatum?.users[user.id]

  if (Object.keys(showDatum.users).length) {
    await updateDoc(`shows/${id}`, showDatum.users)
  } else {
    await deleteDoc(`shows/${id}`)
  }
}
