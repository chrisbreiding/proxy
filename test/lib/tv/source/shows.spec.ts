import nock from 'nock'
import { beforeEach, describe, expect, it, vi } from 'vitest'

const apikey = process.env.THETVDB_API_KEY = 'api-key'
const pin = process.env.THETVDB_PIN = 'pin'

import { startServer } from '../../../../index'
import { getShowsUpdatedSince } from '../../../../lib/tv/source/shows'
import { baseUrl } from '../../../../lib/tv/source/util'
import { getDocWhere } from '../../../../lib/tv/store/firebase'
import { fixtureContents, handleServer } from '../../../util'
import { nockLogin } from '../util'

vi.mock('../../../../lib/tv/store/firebase', () => {
  return {
    getDocWhere: vi.fn(),
  }
})

describe('lib/tv/source/shows', () => {
  beforeEach(() => {
    nockLogin(apikey, pin)
  })

  describe('GET /tv/shows/search', () => {
    handleServer(startServer)

    beforeEach(() => {
      vi.clearAllMocks()
      // @ts-ignore
      getDocWhere.mockResolvedValue({ id: 'user-1' })
    })

    it('returns shows found in search', async (ctx) => {
      nock(baseUrl)
      .matchHeader('Authorization', 'Bearer token')
      .get('/v4/search?query=Breaking+Bad&type=series')
      .reply(200, fixtureContents('tv/search-shows'))

      const res = await ctx.request.get('/tv/shows/search?query=Breaking+Bad')
      .set('api-key', 'user-1-api-key')

      expect(res.status).to.equal(200)
      expect(res.body).toMatchSnapshot()
    })

    it('sends 500 on error', async (ctx) => {
      nock(baseUrl)
      .matchHeader('Authorization', 'Bearer token')
      .get('/v4/search?query=Breaking+Bad&type=series')
      .replyWithError('search failure')

      const res = await ctx.request.get('/tv/shows/search?query=Breaking+Bad')
      .set('api-key', 'user-1-api-key')

      expect(res.status).to.equal(500)
      expect(res.body.error).to.equal('search failure')
    })
  })

  describe('#getShowsUpdatedSince', () => {
    it('returns shows updated since date', async () => {
      nock(baseUrl)
      .matchHeader('Authorization', 'Bearer token')
      .get('/v4/updates?action=update&type=series&since=1667707200')
      .reply(200, fixtureContents('tv/updated-shows'))

      const result = await getShowsUpdatedSince('2022-11-06')

      expect(result).toMatchSnapshot()
    })
  })
})
