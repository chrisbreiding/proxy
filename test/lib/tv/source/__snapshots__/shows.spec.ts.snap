// Vitest Snapshot v1

exports[`lib/tv/source/shows > #getShowsUpdatedSince > returns shows updated since date 1`] = `
[
  {
    "dateTime": "2022-11-06T04:00:58.000Z",
    "id": "83498",
  },
  {
    "dateTime": "2022-11-06T04:01:34.000Z",
    "id": "425416",
  },
  {
    "dateTime": "2022-11-06T04:01:35.000Z",
    "id": "76703",
  },
  {
    "dateTime": "2022-11-06T04:02:34.000Z",
    "id": "365414",
  },
  {
    "dateTime": "2022-11-06T04:03:27.000Z",
    "id": "357741",
  },
]
`;

exports[`lib/tv/source/shows > GET /tv/shows/search > returns shows found in search 1`] = `
[
  {
    "description": "When Walter White, a chemistry teacher, is...",
    "firstAired": "2008-01-20T05:00:00.000Z",
    "id": "81189",
    "name": "Breaking Bad",
    "network": "AMC",
    "poster": "https://artworks.thetvdb.com/banners/posters/81189-10.jpg",
    "status": "Ended",
  },
  {
    "description": "明朝时，皇帝出征新疆。。。",
    "firstAired": "2016-04-11T04:00:00.000Z",
    "id": "415491",
    "name": "绝命卦师",
    "poster": "https://artworks.thetvdb.com/banners/v4/series/415491/posters/61f834ad1d22d.jpg",
    "status": "Ended",
  },
  {
    "description": "Metástasis Se centra en Walter Blanco ...",
    "firstAired": "2013-06-23T04:00:00.000Z",
    "id": "273859",
    "name": "Metástasis",
    "network": "Univision",
    "poster": "https://artworks.thetvdb.com/banners/posters/273859-1.jpg",
    "status": "Ended",
  },
  {
    "id": "357407",
    "name": "Baking Bad",
    "network": "Twitch",
    "poster": "https://artworks.thetvdb.com/banners/images/missing/series.jpg",
    "status": "Ended",
  },
  {
    "description": "Politics, Movies, Television, and Music Videos...",
    "firstAired": "2011-03-21T04:00:00.000Z",
    "id": "287759",
    "name": "Bad Lip Reading",
    "network": "YouTube",
    "poster": "https://artworks.thetvdb.com/banners/posters/287759-1.jpg",
    "status": "Continuing",
  },
]
`;
